graphics/src/SubMesh.cc

```diff
@@ -573,10 +573,56 @@ void SubMesh::FillArrays(double **_vertArr, int **_indArr) const
   }
 }
 
+namespace {
+// Simple way to find neighbors by grouping all vertices
+// by X coordinate with (ordered) map. KD-tree maybe better
+// but not sure about construction overhead
+struct Neighbors
+{
+  Neighbors(const std::vector<unsigned int> &_indices,
+            const std::vector<gz::math::Vector3d> &_vertices)
+    : vertices(_vertices)
+  {
+    for (unsigned int i = 0; i < _indices.size(); ++i)
+    {
+      const auto index = _indices[i];
+      this->neighbors[_vertices[index].X()].push_back(index);
+    }
+  }
+
+  template<typename Visitor>
+  void Visit(const gz::math::Vector3d &_point, Visitor _v) const
```

Comment:
can you add some documentation to explain the logic in this function? I think it's to find and include all indices of vertices that are within the tolerance of the `math::equal` checks, so that their normals all count towards the final calculation?